### PR TITLE
Fixes byte order of encapsulated pdu

### PR DESCRIPTION
**Pull request by mroethke**
and adds a test for this bug.

### CODE REVIEW

**Comment by waehlisch** [comment]
@shaddeus would you like to rive a review, because this relates to your isseu #87?

**Comment by shaddeus** [comment]
@mroethke, nice! @waehlisch the fix isn't trivial, I'll look at the commits on the weekend.

**Comment by waehlisch** [comment]
Ping @shaddeus

File: rtrlib/rtr/packets.c [hunk 1]
```diff
@@ -192,102 +209,136 @@ void rtr_change_socket_state(struct rtr_socket *rtr_socket, const enum rtr_socke
...
-    struct pdu_header *header = pdu;
-    struct pdu_error * err_pdu = NULL;
+static void rtr_pdu_convert_footer_byte_order(void *pdu, enum target_byte_order target_byte_order) {
+	uint32_t (*convert_long)(uint32_t);
+	struct pdu_error *err_pdu;
+	struct pdu_header *header = pdu;
+	uint32_t addr6[4];
+	const enum pdu_type type = rtr_get_pdu_type(pdu);
+
+	if (target_byte_order == TO_NETWORK_BYTE_ORDER) {
```
Comment (shaddeus) [inline rtrlib/rtr/packets.c]: This code is duplicate with line 217 in function `rtr_pdu_convert_header_byte_order()`. Consider `get_convert_short_fn` and `get_convert_long_fn` as static functions with parameter `enum target_byte_order` and returning pointer to functions `htons`/`ntohs` or `htonl`/`ntohl`

File: rtrlib/rtr/packets.c [hunk 2]
```diff
@@ -192,102 +209,136 @@ void rtr_change_socket_state(struct rtr_socket *rtr_socket, const enum rtr_socke
...
+		err_pdu = (struct pdu_error *) pdu;
+		if (target_byte_order == TO_NETWORK_BYTE_ORDER) {
+			*((uint32_t *)(err_pdu->rest + err_pdu->len_enc_pdu)) =
+			    convert_long(*((uint32_t *)(err_pdu->rest + err_pdu->len_enc_pdu)));
+			err_pdu->len_enc_pdu = convert_long(err_pdu->len_enc_pdu);
+		} else {
+			err_pdu->len_enc_pdu = convert_long(err_pdu->len_enc_pdu);
+			*((uint32_t *)(err_pdu->rest + err_pdu->len_enc_pdu)) =
+			    convert_long(*((uint32_t *)(err_pdu->rest + err_pdu->len_enc_pdu)));
+		}
```
Comment (shaddeus) [inline rtrlib/rtr/packets.c]: Great you have noticed the different sequence of lines.

**Comment by waehlisch** [comment]
Great, thanks @shaddeus!
@mroethke can you update the code and consider the comments?

**Comment by mroethke** [comment]
Done, please take another look.

File: rtrlib/lib/ipv6.h [hunk 3]
```diff
@@ -61,11 +61,14 @@ int lrtr_ipv6_str_to_addr(const char *a, struct lrtr_ipv6_addr *ip);
 
 /**
  * @ingroup util_h[{
- * Converts the passed IPv6 address first_bit network byte order to host byte order.
- * @param[in] src Pointer to a uint32_t array storing a ipv6 address in network byte order.
- * @param[out] dest Pointer a uint32_t array that will be used to store the ipv6 addr in host byte order.
+ * Converts the passed IPv6 address byte order with the given function.
+ * @param[in] src Pointer to a uint32_t array storing a ipv6 address in source byte order.
+ * @param[out] dest Pointer a uint32_t array that will be used to store the ipv6 addr in target byte order.
+ * @param[in] convert_fp Pointer to the function used vor conversion
```
Comment (smlng) [inline rtrlib/lib/ipv6.h]: /vor/for/

File: rtrlib/rtr/packets.c [hunk 4]
```diff
@@ -1130,8 +1191,9 @@ int rtr_send_error_pdu(const struct rtr_socket *rtr_socket, const void *erroneou
 	err_pdu->len = msg_size;
 
 	err_pdu->len_enc_pdu = erroneous_pdu_len;
-	if (erroneous_pdu_len > 0)
+	if (erroneous_pdu_len > 0) {
```
Comment (smlng) [inline rtrlib/rtr/packets.c]: IIRC this braces are not allowed in kernel coding style:

> Do not unnecessarily use braces where a single statement will do.

As shown 5 lines below

File: rtrlib/rtr/packets.c [hunk 5]
```diff
@@ -1140,6 +1202,45 @@ int rtr_send_error_pdu(const struct rtr_socket *rtr_socket, const void *erroneou
...
+                                        const void *erroneous_pdu,
+                                        const uint32_t erroneous_pdu_len,
+                                        const enum pdu_error_type error,
+                                        const char *err_text,
+                                        const uint32_t err_text_len)
+{
+	char pdu[erroneous_pdu_len];
+	memcpy(&pdu, erroneous_pdu, erroneous_pdu_len);
+
+	if (erroneous_pdu_len == sizeof(struct pdu_header)) {
```
Comment (smlng) [inline rtrlib/rtr/packets.c]: braces not needed here, all 1 line statements

**Comment by mroethke** [comment]
spelling fixed and unnecessary braces removed.

**Comment by waehlisch** [comment]
@shaddeus final ACK from your side?

**Comment by waehlisch** [comment]
Thanks @shaddeus and @smlng for the review.

**Comment by smlng** [comment]
same here, closed without merge? Hit the wrong button?

**Pull request merged**